zio-http/src/main/scala/zhttp/api/RequestParser.scala

```diff
@@ -0,0 +1,333 @@
+package zhttp.api
+
+import zhttp.http.Request
+import zio.schema.Schema
+
+/**
+ * A RequestParser is a description of a Path, Query Parameters, and Headers.:
+ *   - Path: /users/:id/posts
+ *   - Query Parameters: ?page=1&limit=10
+ *   - Headers: X-User-Id: 1 or Accept: application/json
+ */
+sealed trait RequestParser[A] extends Product with Serializable { self =>
+  private[api] def ++[B](that: RequestParser[B])(implicit zipper: Zipper[A, B]): RequestParser[zipper.Out] =
+    RequestParser.ZipWith[A, B, zipper.Out](self, that, zipper.zip(_, _), zipper.unzip)
+
+  def map[B](f: A => B)(g: B => A): RequestParser[B] =
+    RequestParser.Map(self, f, g)
+
+  private[api] def getPath: Path[_] = {
+    def getPathImpl(requestParser: RequestParser[_]): Option[Path[_]] =
+      requestParser match {
+        case RequestParser.ZipWith(left, right, _, _) =>
+          getPathImpl(left) orElse getPathImpl(right)
+        case RequestParser.Map(info, _, _)            =>
+          getPathImpl(info)
+        case _: Header[_]                             =>
+          None
+        case _: Query[_]                              =>
+          None
+        case route: Path[_]                           =>
+          Some(route)
+      }
+
+    getPathImpl(self).get
+  }
+
+  private[api] def getQueryParams: Option[Query[_]] =
+    self match {
+      case zip: RequestParser.ZipWith[_, _, _] =>
+        (zip.left.getQueryParams, zip.right.getQueryParams) match {
+          case (Some(left), Some(right)) => Some(left ++ right)
+          case (Some(left), None)        => Some(left)
+          case (None, Some(right))       => Some(right)
+          case (None, None)              => None
+        }
+      case RequestParser.Map(info, _, _)       =>
+        info.getQueryParams
+      case route: Query[_]                     =>
+        Some(route)
+      case _: Header[_]                        =>
+        None
+      case _: Path[_]                          =>
+        None
+    }
+
+  private[api] def getHeaders: Option[Header[_]] =
+    self match {
+      case zip: RequestParser.ZipWith[_, _, _] =>
+        (zip.left.getHeaders, zip.right.getHeaders) match {
+          case (Some(left), Some(right)) => Some(left ++ right)
+          case (Some(left), None)        => Some(left)
+          case (None, Some(right))       => Some(right)
+          case (None, None)              => None
+        }
+      case RequestParser.Map(info, _, _)       =>
+        info.getHeaders
+      case route: Header[_]                    =>
+        Some(route)
+      case _: Query[_]                         =>
+        None
+      case _: Path[_]                          =>
+        None
+    }
+
+  private[api] def parseRequest(request: Request): Option[A] = Option(parseRequestImpl(request))
+
+  private[api] def parseRequestImpl(request: Request): A
```

Comment:
```suggestion
  private[api] def parseRequestImpl(request: HttpRequest): A
```

We can improve performance by using the netty request. That will further improve performance.

Reply:
Not sure how to get the netty HttpRequest to where I need it. Might we need to add a variant to Http that allows direct access to the netty Request?